### ISSUE TITLE
[unbound-method] should ignore assignments

### ISSUE THREAD

**Issue by andreineculau** [typescript-eslint/typescript-eslint#1256]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/unbound-method": "error"
  }
}
```

```TS
// PLEASE IGNORE, BAD REPRO
// let obj = {};
// obj.fn = (function() {}).bind(obj);

class A {
  foo(): void {
    this.foo = undefined;
  }
}
```

**Expected Result**

No error.

**Actual Result**

'unbound-method' complains about ~`obj.fn`~ `this.foo`.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.7.0` |
| `@typescript-eslint/parser`        | `2.7.0` |
| `TypeScript`                       | `3.7.2` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `12.12.0` |
| `npm`                              | `6.11.3` |

**Comment by bradzacher** [comment]
Unable to repro this against master

![image](https://user-images.githubusercontent.com/7462525/69502630-2f69d580-0ec6-11ea-9cfd-b22be0149041.png)

**Comment by andreineculau** [comment]
Brad, sorry for the bad repro! I updated the example.

<img width="805" alt="Screen Shot 2019-11-25 at 00 14 58" src="https://user-images.githubusercontent.com/708161/69503306-c2336000-0f18-11ea-9bf5-4ebcbbba449f.png">

**Comment by bradzacher** [comment]
Yup, that repros.

Would recommend not assigning to method definitions - generally people treat method definitions as `readonly`.

If you want to reassign, would recommend converting it to a class property.

Happy to accept a PR to add support for assignments.

**Comment by FlorianWendelborn** [comment]
Here’s a very common use-case that triggers this for me:

```tsx
class MyComponent {
	public static defaultProps = {
		name: 'example',
	}

	constructor(props) {
		super(props)
		this.handleClick = this.handleClick.bind(this)
	}

	public render() {
		return <div onClick={this.handleClick}>Click Me</div>
	}

	private handleClick() {
		const { name, onClick } = this.props
		onClick(name)
	}
}
```

**Comment by PachowStudios** [comment]
This is also causing problems in my use case:

```ts
BigNumber.prototype.valueOf = () => {
// ^^^^^^^^^^^^^^^^^^ eslint(@typescript-eslint/unbound-method)
  throw new Error(
    'Implicit conversion of BigNumber to primitive type is prohibited. Explicitly use toString() instead.',
  );
};
```

**Comment by bradzacher** [comment]
this project is run by volunteers, and relies upon community contributions to move forward.
Happy to accept a PR to add support for assignments.

**Comment by danvim** [comment]
https://stackoverflow.com/questions/60469197/typescript-nextpage-function-getinitialprop-static-method-with-typescript-eslin

I just found this issue. I asked this question on SO recently and this seems to be related. So I'm just posting it here as well.
